mm2src/coins/eth.rs

```diff
@@ -1243,55 +1245,348 @@ impl SwapOps for EthCoin {
+
+        let fut = async move {
+            let tx_from_rpc = selfi
+                .web3
+                .eth()
+                .transaction(TransactionId::Hash(H256::from(taker_fee_hash.as_slice())))
+                .compat()
+                .await
+                .map_to_mm(|e| ValidatePaymentError::InvalidRpcResponse(e.to_string()))?;
+
+            let tx_from_rpc = match tx_from_rpc {
+                Some(t) => t,
+                None => {
+                    return MmError::err(ValidatePaymentError::WrongPaymentTx(format!(
+                        "Didn't find provided tx {:?} on ETH node",
+                        tx_from_rpc
+                    )))
+                },
+            };
+
+            if tx_from_rpc.from != sender_addr {
+                return MmError::err(ValidatePaymentError::WrongPaymentTx(format!(
+                    "{}: Fee tx {:?} was sent from wrong address, expected {:?}",
+                    INVALID_SENDER_ERR_LOG, tx_from_rpc, sender_addr
+                )));
+            }
+
+            if let Some(block_number) = tx_from_rpc.block_number {
+                if block_number <= min_block_number.into() {
+                    return MmError::err(ValidatePaymentError::WrongPaymentTx(format!(
+                        "{}: Fee tx {:?} confirmed before min_block {}",
+                        EARLY_CONFIRMATION_ERR_LOG, tx_from_rpc, min_block_number
+                    )));
+                }
+            }
+
+            //TODO: Validate if taker fee is old
+
+            match &selfi.coin_type {
+                EthCoinType::Eth => {
+                    if tx_from_rpc.to != Some(fee_addr) {
+                        return MmError::err(ValidatePaymentError::WrongPaymentTx(format!(
+                            "{}: Fee tx {:?} was sent to wrong address, expected {:?}",
+                            INVALID_RECEIVER_ERR_LOG, tx_from_rpc, fee_addr
+                        )));
+                    }
+                },
+                EthCoinType::Erc20 {
+                    platform: _,
+                    token_addr,
+                } => {
+                    if tx_from_rpc.to != Some(*token_addr) {
+                        return MmError::err(ValidatePaymentError::WrongPaymentTx(format!(
+                            "{}: ERC20 Fee tx {:?} called wrong smart contract, expected {:?}",
+                            INVALID_CONTRACT_ADDRESS_ERR_LOG, tx_from_rpc, token_addr
+                        )));
+                    }
+
+                    let function = ERC20_CONTRACT
+                        .function("transfer")
+                        .map_to_mm(|e| ValidatePaymentError::InternalError(e.to_string()))?;
+                    let decoded_input = function
+                        .decode_input(&tx_from_rpc.input.0)
+                        .map_to_mm(|e| ValidatePaymentError::TxDeserializationError(e.to_string()))?;
+
+                    if decoded_input[0] != Token::Address(fee_addr) {
```

Comment:
Direct access memory addresses might cause problems which are hard to debug. I recommend to use `first()` function. It will return `None` rather than panicking the mm2.